packages/remirror__extension-link/src/link-extension.ts

```diff
@@ -739,6 +835,91 @@ export class LinkExtension extends MarkExtension<LinkOptions> {
 
     return autoLinkAllowedTLDs.includes(tld);
   }
+
+  private findURL(input: string): string | undefined {
+    for (const match of findMatches(input, this.options.autoLinkRegex)) {
+      const urlMatch = getMatchString(match);
+
+      if (!this.isValidUrl(urlMatch)) {
+        continue;
+      }
+
+      return urlMatch;
+    }
+
+    return;
+  }
+
+  private getURLEndIndex({
+    url,
+    startIndex = 0,
+    text,
+  }: {
+    url?: string;
+    startIndex?: number;
+    text: string;
+  }) {
+    const newUrl = createNewURL(text);
+
+    // Check for trailing punctuation and adjacent characters
+    if (newUrl) {
+      const endsWithPunctuation = (input: string) =>
+        this.options.adjacentPunctuations.includes(input.slice(-1));
+
+      const adjacentCharCount = getTrailingCharCount({
+        input: text,
+        url,
+      });
+
+      const { pathname, search } = newUrl;
+
+      let decoded;
+
+      try {
+        decoded = decodeURI(pathname.slice(1) + search);
+      } catch {
+        decoded = pathname.slice(1) + search;
+      }
+
+      if (decoded && endsWithPunctuation(decoded)) {
+        const index = -1;
+        const balancedIndex = !isBalanced(decoded) ? getBalancedIndex(decoded, index) : 0;
+        const trailingPunctuationIndex = SENTENCE_PUNCTUATIONS.includes(decoded.slice(-1))
```

Comment:
No need to calculate this until line 900

Reply:
good catch